compiler/rustc_codegen_ssa/src/mir/mod.rs

```diff
@@ -83,8 +83,8 @@ pub struct FunctionCx<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> {
     /// Cached unreachable block
     unreachable_block: Option<Bx::BasicBlock>,
 
-    /// Cached terminate upon unwinding block
-    terminate_block: Option<Bx::BasicBlock>,
+    /// Cached terminate upon unwinding (reason: InCleanup) block
+    terminate_in_cleanup_block: Option<Bx::BasicBlock>,
```

Comment:
Here (and in `mir/patch.rs`) I had to guess whether we should rather cache a block for the InCleanup reason or the Abi reason. I hope this is the right guess.^^ Or should we have a cache for each reason?

Reply:
This cache was introduced in #102906 and there wasn't any additional commentary on its purpose - presumably just to improve performance. Given that, caching both reasons probably makes sense - maybe `Option<(Bx::BasicBlock, UnwindTerminateReason)>` would be sufficient?

Reply:
Or use one field for each terminate reason.

Reply:
> > Or use one field for each terminate reason.
> 
> That sounds terrible, we might want to extend this enum in the future.

We are already caching an unreachable block too.

Reply:
>      We are already caching an unreachable block too.

Sure, but that's a pretty different kind of block at least. I'd rather treat the `reason` uniformly wherever possible. Your proposal would require adding a match on the reason and dispatching to the right field at the beginning and end of each of these functions that have a cache... that's really unpleasant IMO.